pkg/hub/managedclusterset/rootclustersetcontroller.go

```diff
@@ -0,0 +1,167 @@
+package managedclusterset
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	clientset "open-cluster-management.io/api/client/cluster/clientset/versioned"
+	clusterinformerv1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterinformerv1beta1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1beta1"
+	clusterlisterv1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	clusterlisterv1beta1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1beta1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+//go:embed manifests
+var manifestFiles embed.FS
+
+const rootManagedClusterSetName = "RootManagedClusterSet"
+
+type rootManagedClusterSetController struct {
+	kubeClient       kubernetes.Interface
+	clusterClient    clientset.Interface
+	clusterLister    clusterlisterv1.ManagedClusterLister
+	clusterSetLister clusterlisterv1beta1.ManagedClusterSetLister
+	cache            resourceapply.ResourceCache
+	eventRecorder    events.Recorder
+
+	// rootClusterSet caches the mappings between clusters and clustersets.
+	// With the mappings, it's easy to find out which clustersets are impacted on a
+	// change of cluster.
+	rootClusterSet sets.String
+
+	// mapLock protects the read/write of rootClusterSet
+	mapLock sync.RWMutex
+}
+
+func NewRootManagedClusterSetController(
+	kubeClient kubernetes.Interface,
+	clusterClient clientset.Interface,
+	clusterInformer clusterinformerv1.ManagedClusterInformer,
+	clusterSetInformer clusterinformerv1beta1.ManagedClusterSetInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &rootManagedClusterSetController{
+		kubeClient:       kubeClient,
+		clusterClient:    clusterClient,
+		clusterLister:    clusterInformer.Lister(),
+		clusterSetLister: clusterSetInformer.Lister(),
+		cache:            resourceapply.NewResourceCache(),
+		eventRecorder:    recorder.WithComponentSuffix("root-managed-cluster-set-controller"),
+		rootClusterSet:   sets.NewString(),
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
```

Comment:
Use `WithFilteredEventsInformersQueueKeyFunc` instead of `WithInformersQueueKeyFunc` to add a filter function. The controller only cares about the particular clusterset.